src/MassTransit.Abstractions/Internals/Extensions/TaskExtensions.cs

```diff
@@ -105,17 +105,32 @@ async Task<T> WaitAsync()
 
             async Task WaitAsync()
             {
-                var delayTask = Task.Delay(Debugger.IsAttached ? Timeout.InfiniteTimeSpan : timeout, cancellationToken);
-
-                var completed = await Task.WhenAny(task, delayTask).ConfigureAwait(false);
-                if (completed == delayTask)
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                try
                 {
-                    task.IgnoreUnobservedExceptions();
+                    Task delayTask = Task.Delay(Debugger.IsAttached ? Timeout.InfiniteTimeSpan : timeout, cts.Token)
+                        .ContinueWith(t => t, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
 
-                    throw new TimeoutException(FormatTimeoutMessage(memberName, filePath, lineNumber));
-                }
+                    var completed = await Task.WhenAny(task, delayTask).ConfigureAwait(false);
 
-                task.GetAwaiter().GetResult();
+                    cts.Cancel();
+
+                    if (completed == delayTask)
```

Comment:
No need to cancel the delay task if it's already completed.
```suggestion
                    if (completed == delayTask)
                    {
                        // ...
                    }
                    else
                    {
                        cts.Cancel();
                    }
```

Reply:
agree on this